network/src/transport.rs

```diff
@@ -207,193 +218,948 @@ pub async fn perform_handshake<T: TSocket>(
+    addr: NetworkAddress,
+    remote_pubkey: x25519::PublicKey,
+) -> io::Result<Connection<NoiseStream<T>>> {
+    let origin = ConnectionOrigin::Outbound;
+    let socket = fut_socket.await?;
+
+    // we can check early if this is even a trusted peer
+    let peer_id = identity_pubkey_to_peer_id(ctxt.trusted_peers.as_ref(), &remote_pubkey)?;
+
+    // try authenticating via noise handshake
+    let is_mutual_auth = ctxt.trusted_peers.is_some();
+    let socket = ctxt
+        .noise
+        .upgrade_outbound(socket, is_mutual_auth, remote_pubkey)
+        .await?;
+
+    // sanity check: Noise IK should always guarantee this is true
+    debug_assert_eq!(remote_pubkey, socket.get_remote_static());
+
+    // try to negotiate common libranet version and supported application protocols
+    perform_handshake(peer_id, socket, addr, origin, &ctxt.own_handshake).await
+}
+
+/// The common LibraNet Transport.
+///
+/// The base transport layer is pluggable, so long as it provides a reliable,
+/// ordered, connection-oriented, byte-stream abstraction (e.g., TCP). We currently
+/// use either `MemoryTransport` or `TcpTransport` as this base layer.
+///
+/// Inbound and outbound connections are first established with the `base_transport`
+/// and then negotiate a secure, authenticated transport layer (currently Noise
+/// protocol). Finally, we negotiate common supported application protocols with
+/// the `Handshake` protocol.
+// TODO(philiphayes): rework Transport trait, possibly include Upgrade trait.
+// ideas in this PR thread: https://github.com/libra/libra/pull/3478#issuecomment-617385633
+pub struct LibraNetTransport<TTransport> {
+    base_transport: TTransport,
+    ctxt: Arc<UpgradeContext>,
+    identity_pubkey: x25519::PublicKey,
+}
+
+impl<TTransport> LibraNetTransport<TTransport>
+where
+    TTransport: Transport<Error = io::Error>,
+    TTransport::Output: TSocket,
+    TTransport::Outbound: Send + 'static,
+    TTransport::Inbound: Send + 'static,
+    TTransport::Listener: Send + 'static,
+{
+    pub fn new(
+        base_transport: TTransport,
+        identity_key: x25519::PrivateKey,
+        trusted_peers: Option<Arc<RwLock<HashMap<PeerId, NetworkPublicKeys>>>>,
+        handshake_version: u8,
+        application_protocols: SupportedProtocols,
+    ) -> Self {
+        let mut own_handshake = HandshakeMsg::new();
+        own_handshake.add(SUPPORTED_MESSAGING_PROTOCOL, application_protocols);
+        let identity_pubkey = identity_key.public_key();
+
+        // Only use anti replay protection in mutual-auth scenarios. In theory,
+        // this is applicable everywhere; however, we would need to spend some
+        // time making this more sophisticated so it garbage collects old
+        // timestamps and doesn't use unbounded space. These are not problems in
+        // mutual-auth scenarios because we have a bounded set of trusted peers
+        // that rarely changes.
```

Comment:
this is not the rational behind not using timestamps in one-way authenticated connections. We don't do it because there is no point for an attacker to replay messages, as they can just create new connections with new public keys (no trusted peers from the server side).

Reply:
If we condense our whole handshake (noise + protocol negotiation) into 1 RTT (which we can actually do today, but we're not rn because it breaks abstractions a bit), replays become much more dangerous, since a successful replay would replace the honest peer's current connection. In this world, anti replay mitigation would help everywhere (unless clients in one-way auth are literally sampling a new static keypair for every connection).

Reply:
ah that's a good point... another way to obtain "key confirmation" here is to wait for a message from the initiator post-handshake. It's a bit ugly but we could in theory add that to the handshake.

This message could be the encryption of an empty message, or even better any negotiation stuff that can happen post-handshake.

Reply:
if we continue as we are today (i.e., second RTT w/ protocol negotiation) then replays are not nearly as dangerous and we can afford to skip the engineering effort required to make the protection work in public-facing endpoints. But that's obviously a pretty implicit property that we should definitely document : )